### PR TITLE
Refactored provider responses in favor of a generic GatewayResponse

### PR DESCRIPTION
**Pull request by ckalista**
## WHAT
- Introduced a generic GatewayResponse encapsulating the BaseResponse returned by a provider. The goal of this new class is to determine whether the BaseResponse represents a success or a fail. In case of a success, GatewayResponse provides easy access to the actual BaseResponse. In case of a fail, GatewayResponse provides access to a GatewayError specifying the error.
- Reworked all provider responses to implement BaseResponse for compliance with GatewayResponse.
- Tidied up *PaymentProvider by moving the logic determining the response status to where it belongs in their respective services.
- Fixed bug forcing the transaction id to be generated and saved prior to issuing a provider request.
- Consolidated unit tests for Authorisation which were missing many scenarios.
- Fixed bug discovered in JAXB mapping of WorldpayOrderStatusResponse, missing the error code and error message during unmarshall inquiry responses from Worldpay.
- Consolidated all Worldpay responses and strenghtened unit tests.
- Added missing integration test in WorldpayRefundITest ensuring the system captures an ERROR record when the gateway operation fails.

### CODE REVIEW

**Comment by simad** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

File: src/main/java/uk/gov/pay/connector/model/gateway/AuthorisationGatewayRequest.java [hunk 1]
```diff
@@ -13,6 +16,10 @@ public AuthorisationGatewayRequest(ChargeEntity charge, Card card) {
         this.card = card;
     }
 
+    public Optional<String> getTransactionId() {
+        return Optional.ofNullable(charge.getGatewayTransactionId());
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/model/gateway/AuthorisationGatewayRequest.java]: Can transactionId be null in any case for a request? if not just remove even the optional

File: src/main/java/uk/gov/pay/connector/service/CancelServiceFunctions.java [hunk 2]
```diff
@@ -48,6 +49,10 @@
                 throw new IllegalStateRuntimeException(reloadedCharge.getExternalId());
             }
             reloadedCharge.setStatus(statusFlow.getLockState());
+
+            logger.info(format("Card cancel request sent - charge_external_id = %s, transaction_id = %s, status = %s",
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/CancelServiceFunctions.java]: Between field and value the equals without spaces as per recommendations. Agreed to be like that for all the elements in logging.

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java [hunk 3]
```diff
@@ -58,29 +58,40 @@ public GatewayResponse doAuthorise(String chargeId, Card cardDetails) {
     @Transactional
     @Override
     public ChargeEntity preOperation(ChargeEntity chargeEntity) {
-        return preOperation(chargeEntity, OperationType.AUTHORISATION, legalStates, ChargeStatus.AUTHORISATION_READY);
+        chargeEntity = preOperation(chargeEntity, OperationType.AUTHORISATION, legalStates, AUTHORISATION_READY);
+        getPaymentProviderFor(chargeEntity).generateTransactionId().ifPresent(chargeEntity::setGatewayTransactionId);
+
+        logger.info(format("Card authorisation request sent - charge_external_id = %s, transaction_id = %s, status = %s",
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java]: Same as said previously

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java [hunk 4]
```diff
@@ -58,29 +58,40 @@ public GatewayResponse doAuthorise(String chargeId, Card cardDetails) {
...
-        reloadedCharge.setStatus(authorisationGatewayResponse.getNewChargeStatus());
-        reloadedCharge.setGatewayTransactionId(authorisationGatewayResponse.getTransactionId());
+
+        ChargeStatus status = operationResponse.getBaseResponse()
+                .map(response -> response.isAuthorised() ? AUTHORISATION_SUCCESS : AUTHORISATION_REJECTED)
+                .orElse(AUTHORISATION_ERROR);
+        String transactionId = operationResponse.getBaseResponse()
+                .map(BaseAuthoriseResponse::getTransactionId).orElse("");
+
+        logger.info(format("Card authorisation response received - charge_external_id = %s, transaction_id = %s, status = %s",
```
Comment (rauligs) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java]: Same here

File: src/test/java/uk/gov/pay/connector/it/resources/ChargeExpiryResourceITest.java [hunk 5]
```diff
@@ -33,10 +33,10 @@ public ChargeExpiryResourceITest() {
     @Test
     public void shouldExpireChargesBeforeAndAfterAuthorisationAndShouldHaveTheRightEvents() {
         String extChargeId1 = addCharge(CREATED, "ref", ZonedDateTime.now().minusHours(1), "gatewayTransactionId1");
-        String extChargeId2 = addCharge(AUTHORISATION_SUCCESS, "ref", ZonedDateTime.now().minusHours(1), "gatewayTransactionId2");
-        String extChargeId3 = addCharge(CAPTURE_SUBMITTED, "ref", ZonedDateTime.now().minusHours(1), "gatewayTransactionId3"); //should not get picked
+        String extChargeId2 = addCharge(AUTHORISATION_SUCCESS, "ref", ZonedDateTime.now().minusHours(1), "transaction-id");
+//        String extChargeId3 = addCharge(CAPTURE_SUBMITTED, "ref", ZonedDateTime.now().minusHours(1), "gatewayTransactionId3"); //should not get picked
```
Comment (rauligs) [inline src/test/java/uk/gov/pay/connector/it/resources/ChargeExpiryResourceITest.java]: Remove comments or uncomment

File: src/test/resources/templates/it/worldpay-cancel-notfication-example-it-dtd-validation-disabled.xml [hunk 6]
```diff
@@ -4,7 +4,7 @@
 <paymentService version="1.4" merchantCode="GBSRECABOFFGDSGBP">
     <reply>
         <ok>
-            <cancelReceived orderCode="{{transactionId}}"/>
+            <cancelReceived orderCode="transaction-id"/>
         </ok>
     </reply>
 </paymentService>
```
Comment (rauligs) [inline src/test/resources/templates/it/worldpay-cancel-notfication-example-it-dtd-validation-disabled.xml]: New line

File: src/test/resources/templates/smartpay/cancel-error-response.xml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<soap:Envelope
+        xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
+        >
+    <soap:Body>
+        <soap:Fault>
+            <faultcode>soap:Server</faultcode>
+            <faultstring>validation 168 Original pspReference required for this operation</faultstring>
+        </soap:Fault>
+    </soap:Body>
+</soap:Envelope>
```
Comment (rauligs) [inline src/test/resources/templates/smartpay/cancel-error-response.xml]: New line

File: src/test/resources/templates/smartpay/capture-error-response.xml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<soap:Envelope
+        xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/"
+        >
+    <soap:Body>
+        <soap:Fault>
+            <faultcode>soap:Server</faultcode>
+            <faultstring>validation 163 Original pspReference required for this operation</faultstring>
+        </soap:Fault>
+    </soap:Body>
+</soap:Envelope>
```
Comment (rauligs) [inline src/test/resources/templates/smartpay/capture-error-response.xml]: New line

File: src/test/resources/templates/worldpay/capture-error-response.xml [hunk 9]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE paymentService PUBLIC "-//WorldPay//DTD WorldPay PaymentService v1//EN"
+        "http://dtd.worldpay.com/paymentService_v1.dtd">
+<paymentService version="1.4" merchantCode="GBSRECABOFFGDSGBP">
+    <reply>
+        <error code="2">
+            <![CDATA[Something went wrong.]]>
+        </error>
+    </reply>
+</paymentService>
```
Comment (rauligs) [inline src/test/resources/templates/worldpay/capture-error-response.xml]: New line

File: src/test/resources/templates/worldpay/refund-error-response.xml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE paymentService PUBLIC "-//WorldPay//DTD WorldPay PaymentService v1//EN"
+        "http://dtd.worldpay.com/paymentService_v1.dtd">
+<paymentService version="1.4" merchantCode="GBSRECABOFFGDSGBP">
+    <reply>
+        <error code="2">
+            <![CDATA[Something went wrong.]]>
+        </error>
+    </reply>
+</paymentService>
```
Comment (rauligs) [inline src/test/resources/templates/worldpay/refund-error-response.xml]: New line

File: src/main/java/uk/gov/pay/connector/model/gateway/AuthorisationGatewayRequest.java (hunk 1 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/model/gateway/AuthorisationGatewayRequest.java]: Well transactionId can indeed be null for some providers like Smartpay.

File: src/main/java/uk/gov/pay/connector/service/CancelServiceFunctions.java (hunk 2 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/CancelServiceFunctions.java]: Done.

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java (hunk 3 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java]: Done.

File: src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java (hunk 4 above)
Comment (ckalista) [inline src/main/java/uk/gov/pay/connector/service/CardAuthoriseService.java]: Done.

File: src/test/java/uk/gov/pay/connector/it/resources/ChargeExpiryResourceITest.java (hunk 5 above)
Comment (ckalista) [inline src/test/java/uk/gov/pay/connector/it/resources/ChargeExpiryResourceITest.java]: Comment removed.

File: src/test/resources/templates/smartpay/cancel-error-response.xml (hunk 7 above)
Comment (ckalista) [inline src/test/resources/templates/smartpay/cancel-error-response.xml]: Done.

File: src/test/resources/templates/smartpay/capture-error-response.xml (hunk 8 above)
Comment (ckalista) [inline src/test/resources/templates/smartpay/capture-error-response.xml]: Done

File: src/test/resources/templates/it/worldpay-cancel-notfication-example-it-dtd-validation-disabled.xml (hunk 6 above)
Comment (ckalista) [inline src/test/resources/templates/it/worldpay-cancel-notfication-example-it-dtd-validation-disabled.xml]: Done

File: src/test/resources/templates/worldpay/capture-error-response.xml (hunk 9 above)
Comment (ckalista) [inline src/test/resources/templates/worldpay/capture-error-response.xml]: Done

File: src/test/resources/templates/worldpay/refund-error-response.xml (hunk 10 above)
Comment (ckalista) [inline src/test/resources/templates/worldpay/refund-error-response.xml]: Done

**Comment by rauligs** [comment]
rebuild

**Comment by rauligs** [comment]
rebuild

**Comment by ckalista** [comment]
rebuild

**Comment by rauligs** [comment]
rebuild

**Pull request merged**